1st approach: add requireResidentKey without adding authenticator selection dictionary

AngeloKai opened the pull request:
As per https://github.com/w3c/webauthn/issues/367, submitting a PR to add key storage enum to ScopedCredentialOptions and add resident key flag to the authenticator data. 

A couple of things about this commit:
- Whether the authenticator exports the private key as the credential ID is entirely up to the authenticator. In fact, many TPMs are designed so that the private key never leaves the device. 
- Instead of "onServer", the value is now "onRelyingParty". The change is made to give flexibility back to the RP as to where it wants to store the key information. While storing on the server is more secure, the RP can store the key in local storage.

vijaybh commented on index.bs:
```diff
@@ -306,6 +311,11 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
+: <dfn>Resident Key</dfn>
+: <dfn>RK</dfn>
+:: A [=resident key=] is a flag representing whether the authenticator is capable of storing <a>credential private key</a> to 
+    generate the [RP]-specific <a>credential key pair</a>. 
```
This definition does not make sense to me. A more natural way to define Resident Key would be as a key that can be looked up and used by the authenticator purely on the basis of its associated RP ID. We should seek to define behavior, not platform implementation details.

The important thing is that the Resident Key is a key, not a flag. If you want to define a flag that indicates an authenticator is capable of creating Resident Keys, then call it the Resident Key Capability Flag or something like that.

vijaybh commented on index.bs:
```diff
@@ -494,17 +504,19 @@ When this method is invoked, the user agent MUST execute the following algorithm
-    1. [=In parallel=], invoke the [=authenticatorMakeCredential=] operation on |authenticator| with |rpId|,
-        |clientDataHash|, {{accountInformation}}, |normalizedParameters|, |excludeList| and |clientExtensions| as parameters.
+    1. [=In parallel=], invoke the <a>authenticatorMakeCredential</a> operation on |authenticator| with |rpId|,
+        |clientDataHash|, {{accountInformation}}, |normalizedParameters|, |excludeList|, |keyStorage| and |clientExtensions| 
```
You haven't defined where |keyStorage| comes from. Did you mean <code>{{options}}.{{ScopedCredentialOptions/keystorage}}</code> ?

vijaybh commented on index.bs:
```diff
@@ -800,6 +813,14 @@ authorizing an authenticator with which to complete the operation.
+        constraint on which authenticators are eligible to participate in a [[#makeCredential]] operation.
+        See [[#attachment]] for a description of the attachment values and their meanings.
+
+    - The <dfn>keyStorage</dfn> parameter contains description of where the <a>credential private key</a> should be stored. 
```
Could we say something like "The <dfn>keyStorage</dfn> parameter indicates the [RP]'s requirements regarding availability of the credential."?

vijaybh commented on index.bs:
```diff
@@ -850,6 +867,27 @@ example of the latter, when the user is accessing the [RP] from a given client f
+
+<pre class="idl">
+    enum KeyStorage {
+        "onAuthenticator",
```
IMO this text is confusing and very likely to be misunderstood. Much better to have choices like say "resident" and "nonresident", and to use text similar to my suggestion on line 317. Otherwise a naive reader of this text might think it is okay to export a private key in the clear and store it on a server.

vijaybh commented on index.bs:
```diff
@@ -1168,7 +1206,9 @@ The [=authenticator data=] structure is a byte array of 37 bytes or more, as fol
             Flags (bit 0 is the least significant bit):
             - Bit 0: [=Test of User Presence=] (`TUP`) result.
-            - Bits 1-5: Reserved for future use (`RFU`).
+            - Bit 1: . 
```
What does this mean?

vijaybh commented on index.bs:
```diff
@@ -1168,7 +1206,9 @@ The [=authenticator data=] structure is a byte array of 37 bytes or more, as fol
             - Bit 0: [=Test of User Presence=] (`TUP`) result.
-            - Bits 1-5: Reserved for future use (`RFU`).
+            - Bit 1: . 
+            - Bit 2: [=Resident Key=] (`RK`) result. 
```
What procedure is this the result of?

vijaybh commented on index.bs:
```diff
@@ -1202,6 +1242,9 @@ requested credential exactly matches the RP ID supplied by the client.
 `RFU` bits in the flags byte SHALL be set to zero.
 
+The `RK` flag SHALL be set if and only if the authenticator is capable of storing <a>credential private key</a> to 
+    generate the [RP]-specific <a>credential key pair</a> on demand. 
```
I'm having difficulty parsing this. Again, why not focus on this specific credential and say that it can be looked up and used in the future by this authenticator based only on RP ID?

vijaybh commented on index.bs:
```diff
@@ -1241,6 +1284,7 @@ input parameters:
     as per the procedure in [[WebCryptoAPI#algorithm-normalization-normalize-an-algorithm]].
 - A list of {{ScopedCredential}} objects provided by the [RP] with the intention that, if any of these are known to the
     authenticator, it should not create a new credential.
+- The {{KeyStorage}} request provided by the [RP]
```
Instead of "request provided by the RP", maybe "option requested by the RP"?

vijaybh commented on index.bs:
```diff
@@ -1257,7 +1301,8 @@ When this operation is invoked, the authenticator must perform the following pro
     - Generate an identifier for this credential, such that this identifier is globally unique with high probability across all
-        credentials with the same type across all authenticators.
+        credentials with the same type across all authenticators. If the {{onRelyingParty}} value is provided by the [RP], 
+        use the <a>credential private key</a> as the identifier for this credential. 
```
No.

vijaybh commented on index.bs:
```diff
@@ -1607,58 +1652,61 @@ ceremony, a [RP] MUST proceed as follows:
+    attestation statement |attStmt|.
+
+1. Store the Credential ID if {{ScopedCredentialOptions/keyStorage/onRelyingParty}} value is supplied in the 
+    <code>{{options}}.{{ScopedCredentialOptions/keyStorage}}</code> parameter. 
```
Please delete this. It is not well-defined. It is not clear what "store" means here, and it is not necessary.

The credential ID must *always* be stored on a successful registration, as specified already in step 13. It does not matter if the credential is resident or not.

AngeloKai commented:
@vijaybh Since last review, I have changed the "key storage" parameter as a Boolean called "requireResidentKey" and defined resident key in the terminology section. In addition, I removed the mention of the parameter in the authenticatorMakeCredential section based on the assumption that the authenticator doesn't need an extra parameter to tell it to store the key. If this isn't true, please let know and I will update that section.

jyasskin commented:
https://github.com/w3c/webauthn/issues/367#issuecomment-283439079 says this PR isn't necessary because the non-resident key is just encrypted into the credential ID that the RP needs to store regardless. Why's @equalsJeffH wrong there?

vijaybh commented:
@jyasskin - AIUI this is seeking to define the difference between different levels of functionality from the authenticator:
- For some authenticators, such as the ones you allude to, the private key material is in the Credential ID. So without a Credential ID, the authenticator is helpless - it cannot provide any assertions. From an RP perspective, this means that they must figure out who the user is first (to some level of confidence) so that they can supply that user's Credential IDs in the allowList to getAssertion.
- For richer authenticators, there is onboard storage. So given only an RP ID, the authenticator can pick a credential (perhaps even with help from the user, if it has UI) and generate an assertion. From the RP's perspective, they can just have the same getAssertion call with an empty allowList for all users, and figure out the username (or more generally, account ID) from the assertion they receive.
The idea is that RPs who only want to build the second type of experience should be able to indicate that.

jyasskin commented:
@vijaybh SG. Maybe that gives a hint as to the right naming for the parameter. It's less about where the key material is stored, and more about whether the Credential ID is a required argument to `getAssertion()`, right? That said, I don't see the Credential ID argument to `getAssertion()` at all, so I'm not sure how the non-storing option ever worked. What am I missing?

AngeloKai commented:
@jyasskin to pass in credential ID, the RP would add a new ScopedCredentialDescriptor with that credential ID in the allowList portion of the AssertionOptions.

jyasskin commented:
@AngeloKai Aha, thanks! In that case, this change should probably make [`ScopedCredentialDescriptor.id`](https://api.csswg.org/bikeshed/?url=https://raw.githubusercontent.com/w3c/webauthn/angelo-keystorage/index.bs#credential-dictionary) non-required so that RPs can avoid passing it when they know the authenticator can find the credential from just their RP ID? (This is #245 but for a different reason.)

AngeloKai commented:
@jyasskin I thought about making cred ID optional before. However, the challenge is that if ID is not required, the allowList as of today is basically useless. I think when Vijay mentioned empty allowList, he meant that the RP doesn't add the allowList parameter in option. If an RP trusts the authenticator to find credentials just based on RP ID, the RP would just not pass in allowList.

vijaybh commented on index.bs:
```diff
@@ -237,7 +237,10 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
+    [=authentication=]. All authenticators refered to this document MUST be capable of the basic operations listed above. 
+    However, certain additive capabilites are only available among certain authenticators, such as the capability to 
+    store <a>credential private keys</a>. When an authenticator is said to be "capable of" performing a task, it means that
+    the authenticator can perform the task when requested. 
```
Editorial nit: I am not sure that last sentence adds anything. Seems like it could be deleted.

vijaybh commented on index.bs:
```diff
@@ -493,18 +501,18 @@ When this method is invoked, the user agent MUST execute the following algorithm
-    <code>{{options}}.{{ScopedCredentialOptions/attachment}}</code> is [=present|not present=] or its value matches
-    |authenticator|'s attachment modality, [=set/append=] |authenticator| to |currentlyAvailableAuthenticators|.
+1. For each |authenticator| currently available on this platform:
+    1. if <code>{{options}}.{{ScopedCredentialOptions/attachment}}</code> is not [=present=] or its value matches |authenticator|'s attachment modality, continue to next step. 
```
Else do what?

vijaybh commented on index.bs:
```diff
@@ -493,18 +501,18 @@ When this method is invoked, the user agent MUST execute the following algorithm
         1. Otherwise, [=list/Append=] |C| to |excludeList|.
-    1. [=In parallel=], invoke the [=authenticatorMakeCredential=] operation on |authenticator| with |rpId|,
-        |clientDataHash|, {{accountInformation}}, |normalizedParameters|, |excludeList| and |clientExtensions| as parameters.
+    1. [=In parallel=], invoke the <a>authenticatorMakeCredential</a> operation on |authenticator| with |rpId|,
```
Why change this markdown?

vijaybh commented on index.bs:
```diff
@@ -800,6 +809,14 @@ authorizing an authenticator with which to complete the operation.
+
+    - The <dfn>requireResidentKey</dfn> parameter describes the [=[RPS]=]' requirements regarding availability of the 
+        <a>credential private key</a>. If the parameter is set to <i>true</i>, the authenticator SHOULD create a <a>resident
+        key</a> when creating a <a>scoped credential</a>. Similiar to {{attachment}}, the {{requireResidentKey}} parameter is an additional constraint on which authenticators are eligible for a [[#makeCredential]] or [[#getAssertion]] operation.
```
Please use the [=term=] style of markup instead of &lt;a>term&lt;/a> to be consistent with the rest of the spec.

vijaybh replied:
It's only a constraint on which authenticators are eligible for makeCredential, right? For getAuthentication, authenticators will either respond or not, there is no reason to set this option.

vijaybh commented on index.bs:
```diff
@@ -1607,58 +1623,58 @@ ceremony, a [RP] MUST proceed as follows:
-    obtain the attestation statement format |fmt|, the [=authenticator data=] |authData|, and the attestation statement
-    |attStmt|.
+1. Perform CBOR decoding on the {{ScopedCredentialInfo/attestationObject}} field of the {{ScopedCredentialInfo}} structure to
+    obtain the Credential ID, the attestation statement format |fmt|, the [=authenticator data=] |authData|, and the 
```
The credential ID is part of authenticator data, why single it out here?

selfissued commented:
It doesn't seem like this is mature enough for the Implementer's Draft.  But since it affects the API, we should probably give the API changes a thumbs-up or thumbs-down before then.

jyasskin commented on index.bs:
```diff
@@ -324,6 +324,11 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
 
+: <dfn>Resident Key</dfn>
+: <dfn>RK</dfn>
+:: A [=resident key=] is a [=credential private keys=] that can be found and used by the authenticator based on only the 
```
"credential private key"

jyasskin replied:
Something in https://w3c.github.io/webauthn/#op-get-assertion is missing a mention of the fact that keys might only be found if their credential IDs are passed in allowList. Without that, the whole idea of resident vs non-resident keys doesn't really make sense.

jyasskin commented on index.bs:
```diff
@@ -551,9 +556,12 @@ When this method is invoked, the user agent MUST execute the following algorithm
-    |authenticator|'s attachment modality, [=set/append=] |authenticator| to |currentlyAvailableAuthenticators|.
+1. For each |authenticator| currently available on this platform:
+    1. If <code>|options|.{{MakeCredentialOptions/attachment}}</code> is [=present|not present=] or its value matches
+        |authenticator|'s attachment modality, continue to next step. Otherwise, do not [=set/append=] |authenticator|
```
"do not X" is odd to see in a spec algorithm. Not-doing things is the default. ;)

Maybe `If <code>|options|.{{MakeCredentialOptions/attachment}}</code> is [=present=] and its value does not match |authenticator|'s attachment modality, [=continue=].`

jyasskin commented on index.bs:
```diff
@@ -551,9 +556,12 @@ When this method is invoked, the user agent MUST execute the following algorithm
+        |authenticator|'s attachment modality, continue to next step. Otherwise, do not [=set/append=] |authenticator|
+        to |currentlyAvailableAuthenticators|.
+    1. If <code>|options|.{{MakeCredentialOptions/requireResidentKey}}</code> flag is not [=present=] or its value
+        is set to [=true=], [=set/append=] |authenticator| to |currentlyAvailableAuthenticators|.
```
Why is present-and-false different from not-present?

jyasskin commented on index.bs:
```diff
@@ -922,6 +931,12 @@ optionally evidence of [=user consent=] to a specific transaction.
 
+    :   <dfn>requireResidentKey</dfn>
+    ::  This member describes the [=[RPS]=]' requirements regarding availability of the [=credential private keys=]. If the
+        parameter is set to <a>true</a>, the authenticator SHOULD create a [=resident key=] when creating a [=scoped credential=].
```
You need to write requirements in an algorithm that actually performs the operations. In this case, it'd be somewhere in authenticatorMakeCredential, not just the non-normative description of the members of MakeCredentialOptions.

I suspect you should s/SHOULD/MUST/. Is there a point in passing `requireResidentKey` if authenticators are still allowed to store the key in the credential ID?

AngeloKai replied:
It is not necessary for that algorithm to be written in this case. We are doing this everywhere in the spec when it comes to authenticator operation. The way the platform finds if the authenticator is available is actually by using the authenticatorGetInfo method, which is documented in the CTAP spec and hard-coded in new authenticators already. The method is not there today anyway because that's really an editorial issue. I am happy with spending sometime later to update that section to more precisely describe the CTAP spec. 
I propose we just pass in the bit for now and update the algorithm to be more precise later. That entire section is already outdated compared to the existing CTAP spec.

jyasskin replied:
This ties into #410 then: if the authenticator model section didn't exist, and merely referred to CTAP, it'd be clear that we need to look there to see the normative effect of `requireResidentKey`.

You should still remove the normative statements from the descriptions of these fields, and just write factual statements about how the CTAP algorithms use their values. (e.g. "the authenticator creates ...") Otherwise, when implementers are reading the algorithm definitions to figure out what to implement, they're likely to miss the extra requirements we place in the field descriptions, leading to bugs.

equalsJeffH commented:
I just merged master into the angelo-keystorage branch to facilitate reviewing. 

Also, these linking errors appear when building the spec:
```
> curl https://api.csswg.org/bikeshed/ -F file=@index.bs -F force=1 -F output=err
\033[1;33mLINK ERROR:\033[0m No 'dfn' refs found for 'true'.
&lt;a data-link-type="dfn" data-lt="true">true&lt;/a>
\033[1;33mLINK ERROR:\033[0m No 'dfn' refs found for 'attachment' that are marked for export.
&lt;a data-link-type="dfn" data-lt="attachment">attachment&lt;/a>
\033[1;33mLINK ERROR:\033[0m No 'dfn' refs found for 'requireresidentkey'.
&lt;a data-link-type="dfn" data-lt="requireResidentKey">requireResidentKey&lt;/a>
\033[7;32m ✔ \033[0m Successfully generated, with 3 linking errors
```

AngeloKai commented on index.bs:
```diff
@@ -324,6 +324,11 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
+: <dfn>RK</dfn>
+:: A [=resident key=] is a [=credential private key=] that can be found and used by the authenticator based on only the 
+    associated RP ID. 
+
```
Normally I don't care about renaming. But in this case, I just cannot see any merit to the  name "requireClientSideResidentPrivateKey". Neither the name "ClientSideResidentPrivateKey" nor the name "ResidentKey" means anything to the developer. The developer will read documentation to understand what the method does anyway, just like attachment or timeout. The name should be a hint of the behavior, not a word-by-word description of the behavior of the API. I am perfectly happy with adding more description to the name resident key or add a description that equates "client-side credential private key". In addition, no web developer reads WhatWG/W3C spec. Using fully qualified term such as credential private key or client-side doesn't mean anything to developers. I'd love to spare developers the pain of writing a 35-characters name every time they wants to use a feature.

AngeloKai replied:
I consider the possibility of changing the name to "requireClientResidentKey:. However, most developers' perception of the word "client" is user agent, aka the browser. Using the name "client" may lead to developers thinking the private keys are kept on the browser. Therefore I propose we keep the name "requireResidentKey"

jyasskin replied:
It might help with naming to focus on what this field means for calls to `.get()`. As Vijay [said](https://github.com/w3c/webauthn/pull/378#issuecomment-288467611), if this field is omitted or `false`, the RP must pass a credential ID in [`allowList`](https://w3c.github.io/webauthn/#dom-scopedcredentialrequestoptions-allowlist). If it's `true`, then the RP doesn't need to pass the credential ID. So `mayOmitCredentialId` or `storeCredentialId` could be good names, or if you're willing to invert the default, `willPassCredentialId` or `getRequiresCredentialId`. This is used as `create({publicKey: {storeCredentialId: true}})`, right?

Allowing sites to omit the credential ID also allows the site to rely on browser or authenticator UI to select which account to log into (for a first-factor authenticator), while requiring the site to pass it forces the site to either use the authenticator as a second factor or to have the user explicitly type their username. Given that, maybe `firstFactor` or `authenticatorCanSelectAccount`, although I don't love either of those.

AngeloKai commented on index.bs:
```diff
@@ -563,7 +572,7 @@ When this method is invoked, the user agent MUST execute the following algorithm
     1. [=In parallel=], invoke the [=authenticatorMakeCredential=] operation on |authenticator| with |rpId|,
         |clientDataHash|, |options|.{{MakeCredentialOptions/rp}}, |options|.{{MakeCredentialOptions/user}},
-        |normalizedParameters|, |excludeList| and |clientExtensions| as parameters.
+        |normalizedParameters|, |excludeList|, |clientExtensions|, and |keyStorageDevice| as parameters.
```
It is necessary to pass the bit in to authenticatorMakeCredential given that this option is part of the parameters of the authenticatorMakeCredential operation. keyStorageDevice is the name of that parameter in the CTAP spec.

AngeloKai commented on index.bs:
```diff
@@ -922,6 +932,12 @@ optionally evidence of [=user consent=] to a specific transaction.
+        parameter is set to <a>true</a>, the authenticator SHOULD create a [=resident key=] when creating a [=scoped credential=].
+        Similiar to [=attachment=], the [=requireResidentKey=] parameter is an additional constraint on which authenticators
+        are eligible for {{CredentialsContainer/create()|navigator.credentials.create()}} operation. 
+
```
See above.

equalsJeffH commented:
To be clear, I do not think this essentially ad-hoc approach to authenticator selection is the way to go, regardless of whether the various issues with these particular proposed changes are addressed. 

For example, @leshi is now proposing in issue #424 to also be able to select authenticators by whether or not they are capable of user verification (which is one of the further attributes I listed in [my review above](https://github.com/w3c/webauthn/pull/378#pullrequestreview-34141186)).
